index.js

```diff
@@ -494,13 +537,54 @@ function onStream (client, onerror) {
     }
 
     // All requests to the APM Server must start with a metadata object
+    // The _encodedMetadata property _should_ be set in the Client
+    // constructor function after making a cloud metadata call.
+    //
+    // Since we cork data until the client._encodedMetadata is set the
+    // following conditional should not be necessary. However, we'll
+    // leave it in place out of a healthy sense of caution in case
+    // something unsets _encodedMetadata or _encodedMetadata is somehow
+    // never set.
     if (!client._encodedMetadata) {
       client._encodedMetadata = client._encode({ metadata: client._conf.metadata }, Client.encoding.METADATA)
     }
+
     stream.write(client._encodedMetadata)
   }
 }
 
+/**
+ * Fetches cloud metadata and encodes with other metadata
+ *
+ * This method will encode the fetched cloud-metadata with other metadata
+ * and memoize the data into the _encodedMetadata property.  Data will
+ * be "returned" to the calling function via the passed in callback.
+ *
+ * The cloudMetadataFetcher configuration values is an error first callback
+ * that fetches the cloud metadata.
+ */
+Client.prototype._fetchAndEncodeMetadata = function (cb) {
+  const toEncode = { metadata: this._conf.metadata }
+
+  if (!this._conf.cloudMetadataFetcher) {
+    // no metadata fetcher from the agent -- encode our data and move on
+    this._encodedMetadata = this._encode(toEncode, Client.encoding.METADATA)
+
+    // an "async-hop" to match the async-behavior when `cloudMetadataFetcher` is set
+    process.nextTick(cb, null, this._encodedMetadata)
+  } else {
+    // agent provided a metadata fetcher function.  Call it, use its return
+    // return-via-callback value to set the cloud metadata and then move on
+    this._conf.cloudMetadataFetcher.getCloudMetadata((error, cloudMetadata) => {
```

Comment:
Style nit: This (and many) node packages tend to use the abbreviation `err` for the first error arg in a callback. It would be nice to continue that convention.

Reply:
_nod_ similar feedback from Watson -- I'll bend my polyglot ways to the defacto standard. Fixed